cilium, docker: runtime dependency updates

borkmann opened the pull request:
See commit messages. tl;dr

- Shrinks cilium-runtime from ~150M down to ~97M (!) according to quay.io
  - Final built Cilium container image ~200M smaller (see below)
- Removes generic compiler, linker, binutils and others from image which are not needed anymore
  - Probing fully done by bpftool
- Update to recent clang/llvm for better codegen and remove _all_ backends but BPF
  - Main motivation to reduce verifier complexity due to less stack spilling 
  - Implements BPF CPU probing, allowing also v2 for arm64
- Remove shipping headers (/usr/include/ now empty !) and instead fully rely on Cilium ones
  - Add missing needed headers to bpf/ include infra to be fully independent
- Describe where remaining dependencies are still needed in the Dockerfile

Comparing locally pulled image sizes, tag 2020-02-27 vs 2020-03-20:
```
root@tank:~# docker images
REPOSITORY                      TAG                                        IMAGE ID            CREATED             SIZE
quay.io/cilium/cilium-runtime   2020-03-20                                 3e9d2e373a7f        4 hours ago         229MB
quay.io/cilium/cilium-runtime   2020-02-27                                 33b9abb9ab72        3 weeks ago         433MB
```

```
cilium-new/cilium                 latest                                     697dd743abb9        5 minutes ago       429MB
cilium-old/cilium                 latest                                     ca339d7b069c        About an hour ago   633MB
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10542)
<!-- Reviewable:end -->

borkmann committed:
docker, runtime: remove bpf-map tool in favor of bpftool

bpftool already covers everything we need and if there is a case where
it does not, we should just extend it upstream and pull latest one back
into the image via bpf-next as Dockerfile.runtime does.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
docker, runtime: upgrade to recent clang/llvm image in runtime

Do not rely on clang-7/llvm-7 shipped by Ubuntu base image and
instead upgrade to clang-11/llvm-11 with a BPF-only backend.

This would help overcoming the blockade of [0] where we're hitting
the 1 mio instruction complexity limit which was bisected by Paul
to the kernel commit f7cf25b20 ("bpf: track spill/fill of constants").
As it is stated there:

  Newer clang generates better code by spilling less to the stack.
  Instead it keeps more constants in the registers which hurts
  state pruning since the verifier already tracks constants in the
  registers [...]. Tracking constants in the registers hurts state
  pruning already. Adding tracking of constants through stack hurts
  pruning even more. The later patch address this general constant
  tracking issue with coarse/precise logic.

Side-effect of going with a custom clang-11/llvm-11 build with a
BPF-only backend is that i) we can also shrink the image since x86
is not needed anymore, and ii) avoid shipping a generic compiler in
our image that can generate x86 executable code. This depends on [1]
where we first need to get rid of our custom runtime probes in Cilium
and instead rely on bpftool to take over that job.

Size shrinkage around 36.6M:

  clang-7 (90,777,392) -> clang-11/stripped   (75,617,520)
  llc-7   (51,453,072) -> llc-11/bpf/stripped (29,997,384)

  [0] https://github.com/cilium/cilium/issues/10517
  [1] https://github.com/cilium/cilium/pull/10019

Complexity comparison:

  https://user-images.githubusercontent.com/677393/76440789-aae08b80-63be-11ea-863f-37ab12106ad9.png

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Cc: Paul Chaignon <paul@cilium.io>

borkmann committed:
docker, runtime: remove libc6-dev-i386 dependency

We already got rid of this dependency thanks to recent data path work
in [0], so there is no need to keep installing this dependency in the
docker images. The only place where it's still needed is for building
bpftool therefore move it into runtime-build so switch to gcc-multilib
there.

  [0] https://github.com/cilium/cilium/pull/10204

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

tklauser commented on contrib/packaging/docker/Dockerfile.runtime:
```diff
@@ -8,7 +8,7 @@ apt-get upgrade -y && \
 # Prepackaged Cilium runtime dependencies
 #
 apt-get install -y --no-install-recommends \
-    gpg gpg-agent libelf-dev libmnl-dev libc6-dev-i386 iptables libgcc-5-dev \
```
Just tried this locally to after you reported it. It seems this breaks the `bpftool` build:

```
In file included from skeleton/profiler.bpf.c:4:
In file included from /usr/include/linux/bpf.h:11:
/usr/include/linux/types.h:5:10: fatal error: 'asm/types.h' file not found
#include <asm/types.h>
         ^~~~~~~~~~~~~
1 error generated.
```

I thought that installing `libc-linux-dev:amd64` (which provides said header) would help but for some reason it doesn't.

tklauser replied:
The problem seems to be is that the `/usr/include/asm` symlink is only created once the headers of another arch (in this case for `i386` are installed). For cilium, PR #10204 thus added `-I/usr/include/$(shell uname -m)-linux-gnu` to `$CFLAGS`. But it seems `bpftool` doesn't do that.

borkmann replied:
True, we can move it to the `runtime-build` section though since our BPF code generation does not depend on it anymore. And that said, there's move *-dev stuff we could probably move to there, will check.

borkmann replied:
> The problem seems to be is that the `/usr/include/asm` symlink is only created once the headers of another arch (in this case for `i386` are installed). For cilium, PR #10204 thus added `-I/usr/include/$(shell uname -m)-linux-gnu` to `$CFLAGS`. But it seems `bpftool` doesn't do that.

Could you craft a patch for upstream bpftool on that?

tklauser replied:
Sure, sent https://lore.kernel.org/bpf/20200311123421.3634-1-tklauser@distanz.ch/T/#u

Just noticed that there's a typo in the title: `s/iprofile/profile/`

tklauser reviewed (commented):
`contrib/packaging/deb/Dockerfile` also installs `libc6-dev.i386`. Question is whether this Dockerfile is still in use? It looks quite outdated, e.g. using Go 1.9.

borkmann committed:
docker, runtime: update cilium image runtime tag

Move runtime over to 2020-03-23 for pulling in all base image changes.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

pchaigno commented on contrib/packaging/docker/Dockerfile.runtime:
```diff
@@ -61,6 +49,17 @@ make -j `getconf _NPROCESSORS_ONLN` && \
 #
+# clang/llvm image with only BPF backend
+#
+git clone --depth 1 -b master https://github.com/llvm/llvm-project.git llvm && \
```
@borkmann Did you get a chance to test this on some of the builds affected by #10517 yet? I was under the impression newer versions of Clang actually made things worse. Alexei's commit message reports higher numbers of states processed for newer Clang versions.

borkmann replied:
> @borkmann Did you get a chance to test this on some of the builds affected by #10517 yet? I was under the impression newer versions of Clang actually made things worse. Alexei's commit message reports higher numbers of states processed for newer Clang versions.

I just did a comparison printing all complexity numbers to the kernel log from daemon startup and creation of a single endpoint on same underlying kernel. While clang 7 seems to do better in our init.sh (progs to the left), it seems to generate problematic code for bpf_lxc (progs to the right), and for latest clang-11 the situation is a bit the other way round. This is with mcpu=probe.

![clang 7 vs clang 11 on bpf-next tree](https://user-images.githubusercontent.com/677393/76430600-c8a6f400-63b0-11ea-9989-261910f1e70f.png)

borkmann replied:
Switching to mcpu=v2 overall improves it. New chart with clang 11:

![clang 7 vs clang 11 on bpf-next tree, mcpu=probe_v2](https://user-images.githubusercontent.com/677393/76440789-aae08b80-63be-11ea-863f-37ab12106ad9.png)

aanm commented on Dockerfile.builder:
```diff
@@ -45,9 +45,22 @@ RUN apt-get update \
+	cp bin/clang /usr/bin/clang-git && \
+	cp bin/llc /usr/bin/llc-git && \
+	update-alternatives --install /usr/bin/clang clang /usr/bin/clang-git 100 && \
+	update-alternatives --install /usr/bin/llc llc /usr/bin/llc-git 100
```
I think it would be better to move this section into a dedicated docker image. The clang build seems to take too many resources which might cause the build time to increase unnecessarily. We could build clang separately and have a dedicated image for it as we won't change clang version frequently but we change the image builder frequently.

aanm replied:
Also, we need to update the developer documentation that states that we now depend on this particular clang version.

borkmann replied:
> We could build clang separately and have a dedicated image for it as we won't change clang version frequently but we change the image builder frequently.

Makes sense, was also thinking as another option to leave the builder at clang-7 since it doesn't need to load anything, but otoh, having both versions aligned might be better. I can check once done with the rest. Docs also once the other stuff is resolved in here.

tklauser commented on contrib/packaging/docker/Dockerfile.runtime:
```diff
@@ -37,11 +23,15 @@ RUN \
   gcc git pkg-config bison flex build-essential \
 # Additional bpftool dependencies
-  python3 && \
+  python3 gcc-multilib \
```
Do we still need `gcc-multilib` with the recent changes in the `bpftool` build?

borkmann replied:
Good point, nope. I'm also thinking to shrink down more from the runtime dependencies we install there.

borkmann committed:
docker, runtime: remove gcc-multilib build dependency

Recent work on upstream kernel side, e.g. fe4eb069edb7 ("bpftool: Use
linux/types.h from source tree for profiler build") removed the need
to have gcc-multilib as a bpftool build dependency. Therefore get rid
of it.

Suggested-by: Tobias Klauser <tklauser@distanz.ch>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
docker, runtime: get rid of gpg gpg-agent in runtime image

Digging a bit into the history of this file, 00109a675a00 ("Dockerfile: Incremental
build.") used it for clang 3.7.1. It was actually added as a dependency in commit
de65c8592248 ("docker/Dockerfile: add gpg"). However, since commit 0bf33f653d79
("Update to use LLVM 7."), we don't use it anymore, therefore remove the unneeded
dependency.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
docker, runtime: revisit all *-dev runtime-base dependencies

Move the *-dev packages where possible into the runtime-build section
since for those we need at runtime for BPF compilation, we already
ship headers. Also remove things that are not cruicial for operation
e.g. binutils. The latter is resumably needed for the strip tool, so
move it into runtime-build where we only need it since we don't want
to ship a linker etc in Cilium. Structure the rest a bit to comment
for what a dependency is still needed.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: provide own stddef/stdbool and don't depend on compiler's

Remove the dependency on the system's definition of bool, true,
false and NULL.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

tklauser reviewed (approved):
Nice work digging into all the dependencies and their history!

borkmann committed:
bpf: re-add a proper types.h mapper

Commit a1d93e044c1f ("bpf: don't use fixed size integer types from
stdint.h") removed the types.h mapper to avoid dependency on stdint.h
and instead relied on -I/usr/include/$(shell uname -m)-linux-gnu/ to
pull in linux/types.h from the system headers.

This is problematic in that for the docker runtime base image we want
to get rid of libc6-dev-i386 package which pulls in gcc-multilib and
latter pulls linux-libc-dev which includes all linux headers as well.

Cilium's bpf/ code however already ships all linux header dependencies
by itself, so linux-libc-dev is actually not needed. It's just that we
need to have a proper type mapper for BPF. Implement one with basic
types and also remove the -I/usr/include/$(shell uname -m)-linux-gnu/
inclusion.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
docker, runtime: remove binutils-dev dependency

We currently still have binutils-dev as a runtime dependency for
bpftool since it ships libbfd which is needed for the JIT disasm.
This is not great since it also pulls in binutils itself as a
dependency and as a result the Cilium runtime image ships with an
assembler, linker, etc:

  root@068f3a5ff554:/cni# ls -l /usr/bin/x86_64*
  lrwxrwxrwx 1 root root       7 Jan  8 18:31 /usr/bin/x86_64 -> setarch
  -rwxr-xr-x 1 root root   31152 May  8  2019 /usr/bin/x86_64-linux-gnu-addr2line
  -rwxr-xr-x 1 root root   59640 May  8  2019 /usr/bin/x86_64-linux-gnu-ar
  -rwxr-xr-x 1 root root  917488 May  8  2019 /usr/bin/x86_64-linux-gnu-as
  -rwxr-xr-x 1 root root   26744 May  8  2019 /usr/bin/x86_64-linux-gnu-c++filt
  -rwxr-xr-x 1 root root 2911496 May  8  2019 /usr/bin/x86_64-linux-gnu-dwp
  -rwxr-xr-x 1 root root   31184 May  8  2019 /usr/bin/x86_64-linux-gnu-elfedit
  lrwxrwxrwx 1 root root      24 May  8  2019 /usr/bin/x86_64-linux-gnu-gold -> x86_64-linux-gnu-ld.gold
  -rwxr-xr-x 1 root root  102120 May  8  2019 /usr/bin/x86_64-linux-gnu-gprof
  lrwxrwxrwx 1 root root      23 May  8  2019 /usr/bin/x86_64-linux-gnu-ld -> x86_64-linux-gnu-ld.bfd
  -rwxr-xr-x 1 root root 1779400 May  8  2019 /usr/bin/x86_64-linux-gnu-ld.bfd
  -rwxr-xr-x 1 root root 3136552 May  8  2019 /usr/bin/x86_64-linux-gnu-ld.gold
  -rwxr-xr-x 1 root root   44184 May  8  2019 /usr/bin/x86_64-linux-gnu-nm
  -rwxr-xr-x 1 root root  235720 May  8  2019 /usr/bin/x86_64-linux-gnu-objcopy
  -rwxr-xr-x 1 root root  414128 May  8  2019 /usr/bin/x86_64-linux-gnu-objdump
  -rwxr-xr-x 1 root root   59672 May  8  2019 /usr/bin/x86_64-linux-gnu-ranlib
  -rwxr-xr-x 1 root root  596440 May  8  2019 /usr/bin/x86_64-linux-gnu-readelf
  -rwxr-xr-x 1 root root   31024 May  8  2019 /usr/bin/x86_64-linux-gnu-size
  -rwxr-xr-x 1 root root   31232 May  8  2019 /usr/bin/x86_64-linux-gnu-strings
  -rwxr-xr-x 1 root root  235728 May  8  2019 /usr/bin/x86_64-linux-gnu-strip

We can build bpftool without binutils-dev just fine, it's just that
we won't get the JIT disasm which is an acceptable tradeoff given we
still have the xlated dump and it's not needed often in prod environment.
Also, we can reduce the Cilium runtime image by another ~20M in size.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
vagrant: bump development VM images for clang/llvm

Pull in packer-ci upgrades for clang/llvm so it can run in CI:

  https://github.com/cilium/packer-ci-build/pull/200

For now pull in only the net-next server box since our CI seems to
have massive problems with disk space on nodes. This is currently
being analyzed. We can bump SERVER_VERSION at a later point in time.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: only compile alignchecker with debug information

iproute2's BTF parser otherwise breaks down on newer clang/llvm's
generated BTF, therefore don't generate it for object files we are
later on loading otherwise the daemon will not be able to start up.
We need to find a better fix at some point.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf, cilium: manually probe for v1/v2 BPF CPU availability

There are a couple of downsides using -mcpu=probe right now with the latest
{clang,llc}-11. On latest kernels, -mcpu=probe will select v3 CPU which for
Cilium fails to pass the verifier due to buggy alu32 logic today. I've tried
running with -mcpu=probe -mattr=-alu32 but that doesn't work either at the
moment.

Another downside in LLVM's -mcpu=probe implementation is that anything else
than underlying x86_64 will just always fall back to v1 CPU, sigh:

  StringRef sys::detail::getHostCPUNameForBPF() {
  #if !defined(__linux__) || !defined(__x86_64__)
    return "generic";
  #else
    [...]
  #endif
  }

Given Cilium is in process of being supported on arm64, this is suboptimal.

Therefore, implement a manual probe and select between v1/v2 CPU from the
Cilium agent.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, docs: update versioning note on clang/llvm requirement

State in each relevant section that the latest version is recommended.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: enable tail call on lxc entry when debug is enabled

Otherwise it's increasing complexity further.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
